problem finding installed LLVM on alpine linux

On alpine linux 314, the clang-dev package installs the clang headers to e.g. `usr/include/clang/Frontend/ASTUnit.h` but our scripts end up looking in a path based on `llvm-config --include` e.g. `/usr/lib/llvm11/include/clang/Basic/Version.h`.

I think we could resolve the problem by updating our scripts to look in /usr/include/clang as well / alternatively.

Previous patch ran into an infinite loop. Improved patch (in details below) works better and improves the error message when it is not working.

<details>

```
diff --git a/util/chplenv/chpl_llvm.py b/util/chplenv/chpl_llvm.py
index 80d4e70d..e502fefe 100755
--- a/util/chplenv/chpl_llvm.py
+++ b/util/chplenv/chpl_llvm.py
@@ -88,8 +88,19 @@ def check_llvm_config(llvm_config):
         llvm_header = os.path.join(include_dir,
                                    'llvm', 'Config', 'llvm-config.h')
         llvm_include_ok = os.path.exists(llvm_header)
+        usr_include = "/usr/include"
         clang_header = os.path.join(include_dir, 'clang', 'Basic', 'Version.h')
-        clang_include_ok = os.path.exists(clang_header)
+        clang_header2 = os.path.join(usr_include, 'clang', 'Basic', 'Version.h')
+        if os.path.exists(clang_header):
+            clang_include_ok = True
+        elif os.path.exists(clang_header2):
+            llvmv = run_command([llvm_config, '--version']).strip()
+            clang_ver_f = os.path.join(usr_include,
+                                       'clang', 'Basic', 'Version.inc')
+            with open(clang_ver_f) as f:
+                contents = f.read()
+                if llvmv in contents:
+                    clang_include_ok = True;
 
     s = ''
     if not version_ok:
@@ -113,7 +124,7 @@ def check_llvm_config(llvm_config):
 def find_system_llvm_config():
     llvm_config = overrides.get('CHPL_LLVM_CONFIG', 'none')
     if llvm_config != 'none':
-        return llvm_config
+        return llvm_config, None
 
     homebrew_prefix = chpl_platform.get_homebrew_prefix()
 
@@ -136,27 +147,28 @@ def find_system_llvm_config():
         paths.append(homebrew_prefix + "/opt/llvm/bin/llvm-config")
 
     by_version = defaultdict(list)
-    errs = []
+    first_err = None
 
     for command in paths:
         version, config_err = check_llvm_config(command)
         if not config_err:
             by_version[version].append(command)
-        else:
-            errs.append((command, config_err))
+        elif first_err == None:
+            first_err = config_err
 
     for version in llvm_versions():
         commands = by_version[version]
         if commands:
-            return commands[0]
+            return commands[0], None
 
-    return ''
+    return '', config_err
 
 
 @memoize
 def get_llvm_config():
     llvm_val = get()
     llvm_config = overrides.get('CHPL_LLVM_CONFIG', 'none')
+    llvm_config_err = None
 
     if llvm_val == 'bundled':
         llvm_subdir = get_bundled_llvm_dir()
@@ -166,20 +178,20 @@ def get_llvm_config():
         llvm_config = bundled_config
 
     elif llvm_config == 'none' and llvm_val == 'system':
-        llvm_config = find_system_llvm_config()
+        llvm_config, llvm_config_err = find_system_llvm_config()
 
-    return llvm_config
+    return llvm_config, llvm_config_err
 
 @memoize
 def validate_llvm_config():
     llvm_val = get()
-    llvm_config = get_llvm_config()
+    llvm_config, llvm_config_err = get_llvm_config()
 
     if llvm_val == 'system':
         if llvm_config == '' or llvm_config == 'none':
             error("CHPL_LLVM=system but could not find an installed LLVM"
-                  " with one of the supported versions: {0}".format(
-                  llvm_versions_string()))
+                  " with one of the supported versions: {0}\n{1}".format(
+                  llvm_versions_string(), llvm_config_err))
 
     if (llvm_val == 'system' or
         (llvm_val == 'bundled' and os.path.exists(llvm_config))):
@@ -204,7 +216,7 @@ def validate_llvm_config():
 
 @memoize
 def get_system_llvm_config_bindir():
-    llvm_config = find_system_llvm_config()
+    llvm_config, llvm_config_err = find_system_llvm_config()
     found_version, found_config_err = check_llvm_config(llvm_config)
 
     bindir = None
@@ -231,6 +243,18 @@ def get_system_llvm_clang(lang):
     clang = ''
     if bindir:
         clang = os.path.join(bindir, clang_name)
+        # also try /usr/bin/clang since some OSes use that
+        # for the clang package
+        usrbin = "/usr/bin"
+        clang2 = os.path.join(usrbin, clang_name);
+        if os.path.exists(clang2) and not os.path.exists(clang):
+            llvm_config, llvm_config_err = find_system_llvm_config()
+            # check that clang --version matches llvm-config --version
+            clangv = run_command([clang2, '--version']).strip()
+            llvmv = run_command([llvm_config, '--version']).strip()
+
+            if llvmv in clangv:
+                clang = clang2
 
     return clang
 
@@ -258,7 +282,7 @@ def get_llvm_clang(lang):
 
 
 def has_compatible_installed_llvm():
-    llvm_config = find_system_llvm_config()
+    llvm_config, llvm_config_err = find_system_llvm_config()
 
     if llvm_config:
         clang_c_command = get_system_llvm_clang('c')
@@ -432,7 +456,7 @@ def get_system_llvm_built_sdkroot():
     llvm_val = get()
     host_platform = chpl_platform.get('host')
     if llvm_val == 'system' and host_platform == 'darwin':
-        llvm_config = get_llvm_config()
+        llvm_config, llvm_config_err = get_llvm_config()
         include_dir = run_command([llvm_config, '--includedir']).strip()
         if os.path.isdir(include_dir):
             clang_config = os.path.join(include_dir,
@@ -630,7 +654,7 @@ def get_host_compile_args():
     system = [ ]
 
     llvm_val = get()
-    llvm_config = get_llvm_config()
+    llvm_config, llvm_config_err = get_llvm_config()
 
     if llvm_val == 'system':
         # On Mac OS X with Homebrew, apply a workaround for issue #19217.
@@ -676,7 +700,7 @@ def get_host_link_args():
 
     llvm_dynamic = True
     llvm_val = get()
-    llvm_config = get_llvm_config()
+    llvm_config, llvm_config_err = get_llvm_config()
     clang_static_libs = ['-lclangFrontend',
                          '-lclangSerialization',
                          '-lclangDriver',
@@ -785,7 +809,7 @@ def get_host_link_args():
 
 def _main():
     llvm_val = get()
-    llvm_config = get_llvm_config()
+    llvm_config, llvm_config_err = get_llvm_config()
     llvm_versions = llvm_versions_string()
 
     parser = optparse.OptionParser(usage='usage: %prog [--needs-llvm-runtime]')
diff --git a/util/chplenv/printchplenv.py b/util/chplenv/printchplenv.py
index 4b4207ef..1833e48b 100755
--- a/util/chplenv/printchplenv.py
+++ b/util/chplenv/printchplenv.py
@@ -199,7 +199,8 @@ def compute_all_values():
     ENV_VALS['CHPL_RE2'] = chpl_re2.get()
     ENV_VALS['  CHPL_RE2_IS_OVERRIDDEN'] = chpl_re2.is_overridden()
     ENV_VALS['CHPL_LLVM'] = chpl_llvm.get()
-    ENV_VALS['  CHPL_LLVM_CONFIG'] = chpl_llvm.get_llvm_config()
+    llvm_config, llvm_config_err = chpl_llvm.get_llvm_config()
+    ENV_VALS['  CHPL_LLVM_CONFIG'] = llvm_config
     llvm_clang_c = chpl_llvm.get_llvm_clang('c')
     llvm_clang_cxx = chpl_llvm.get_llvm_clang('c++')
     ENV_VALS['  CHPL_LLVM_CLANG_C'] = " ".join(llvm_clang_c)
```

</details>

But, then it fails with

```
/home/vagrant/chapel/third-party/qthread/qthread-src/src/qloop.c:935:5: error: unknown type name 'ssize_t'; did you mean 'size_t'?
    ssize_t    dynamicBlock;
    ^~~~~~~
    size_t
/usr/include/bits/alltypes.h:50:24: note: 'size_t' declared here
typedef unsigned _Addr size_t;
                       ^
1 error generated.
```

This error appears with the C or LLVM backends selected.

--- a/util/chplenv/chpl_llvm.py
+++ b/util/chplenv/chpl_llvm.py
@@ -96,23 +96,47 @@ def check_llvm_packages(llvm_config):
     clang_cpp_lib = ''
     clang_cpp_lib_ok = False
 
+    usr_include_clang_ok = False
+
     include_dir = run_command([llvm_config, '--includedir']).strip()
     if os.path.isdir(include_dir):
         llvm_header = os.path.join(include_dir,
                                    'llvm', 'Config', 'llvm-config.h')
         llvm_include_ok = os.path.exists(llvm_header)
+
         clang_header = os.path.join(include_dir, 'clang', 'Basic', 'Version.h')
         clang_include_ok = os.path.exists(clang_header)
+        if not clang_include_ok:
+            usr_include = "/usr/include"
+            clang_header2 = os.path.join(usr_include, 'clang', 'Basic', 'Version.h')
+            ver_inc = os.path.join(usr_include, 'clang', 'Basic', 'Version.inc')
+            if os.path.exists(clang_header2) and os.path.exists(ver_inc):
+
+                llvm_version, ignored_err = check_llvm_config(llvm_config)
+                llvm_version = llvm_version.strip()
+                with open(ver_inc) as f:
+                    contents = f.read()
+                    if llvm_version in contents:
+                        usr_include_clang_ok = True
+                        clang_include_ok = True;
 
     llvm_lib_dir = run_command([llvm_config, '--libdir']).strip()
+
+    clang_lib_name = None
+    if sys.platform == "darwin":
+        clang_lib_name = 'libclang-cpp.dylib'
+    elif sys.platform == "win32":
+        clang_lib_name = 'libclang-cpp.dll'
+    else:
+        clang_lib_name = 'libclang-cpp.so'
+
     if os.path.isdir(llvm_lib_dir):
-        if sys.platform == "darwin":
-          clang_cpp_lib = os.path.join(llvm_lib_dir, 'libclang-cpp.dylib')
-        elif sys.platform == "win32":
-          clang_cpp_lib = os.path.join(llvm_lib_dir, 'libclang-cpp.dll')
-        else:
-          clang_cpp_lib = os.path.join(llvm_lib_dir, 'libclang-cpp.so')
+        clang_cpp_lib = os.path.join(llvm_lib_dir, clang_lib_name);
         clang_cpp_lib_ok = os.path.exists(clang_cpp_lib)
+        if usr_include_clang_ok and not clang_cpp_lib_ok:
+            # use e.g. /usr/lib/libclang-cpp.so
+            clang_cpp_lib = os.path.join("/usr/lib", clang_lib_name);
+            clang_cpp_lib_ok = os.path.exists(clang_cpp_lib)
 
     s = ''
     if not llvm_include_ok:
@@ -251,12 +275,27 @@ def get_llvm_clang_command_name(lang):
     else:
         return 'clang'
 
+@memoize
 def get_system_llvm_clang(lang):
     clang_name = get_llvm_clang_command_name(lang)
     bindir = get_system_llvm_config_bindir()
     clang = ''
     if bindir:
         clang = os.path.join(bindir, clang_name)
+
+        if not os.path.exists(clang):
+            # also try /usr/bin/clang since some OSes use that
+            # for the clang package
+            usr_bin = "/usr/bin"
+            clang2 = os.path.join(usr_bin, clang_name);
+            if os.path.exists(clang2):
+                llvm_config = find_system_llvm_config()
+                # check that clang --version matches llvm-config --version
+                clangv = run_command([clang2, '--version']).strip()
+                llvmv = run_command([llvm_config, '--version']).strip()
+
+                if llvmv in clangv:
+                    clang = clang2
 
     return clang
 

